File: ice0208/sliding-window/1004.py
class Solution:
    def longestOnes(self, nums: list[int], k: int) -> int:
        l, r = 0, 0
        cur_zero_cnt = 1 - nums[0]

        answer = 0
        while True:
            if cur_zero_cnt > k:
                l += 1
                cur_zero_cnt -= 1 - nums[l - 1]
                if l > r:
                    r += 1
                    if r == len(nums):
                        break
                    cur_zero_cnt += 1 - nums[r]

            else:
                answer = max(answer, r - l + 1)

                r += 1
                if r == len(nums):
                    break
                cur_zero_cnt += 1 - nums[r]

        return answer


if __name__ == "__main__":
    print(Solution().longestOnes(nums=[1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0], k=2))
    print(
        Solution().longestOnes(
            nums=[0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1], k=3
        )
    )

File: moonpoison/sliding_window/1004.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    int longestOnes(vector<int>& nums, int k) 
    {
        int cnt=0;
        int p1=0, p2=0;
        int ans=0;
        while(p2<nums.size())
        {
            if(nums[p2]==1) p2++; //p2가 1일때 p2증가
            else//p2가 0일때
            {
                if(k>0)//k를 감소하며 p2증가
                {
                    k--;
                    p2++;
                }
                else//감소할 k가 없으면 배열길이계산, p1을 0이있는곳까지 이동후 슬라이딩윈도우
                {
                    ans=max(ans, p2-p1);
                    while(nums[p1]==1) 
                        p1++;    
                    p1++;
                    p2++;
                }
            }
        }
        ans=max(ans, p2-p1);
        return ans;
    }
};